core/src/hir/type_context.rs

```diff
@@ -48,54 +58,872 @@ impl TypeContext {
+        }
+    }
+
+    nodes.map(LifetimeEnv::new)
+}
+
+/// Lowers an [`ast::SelfParam`] into an [`hir::ParamSelf`].
+///
+/// If there are any errors, they're pushed to `errors` and `Err(())` is returned.
+fn lower_self_param(
+    self_param: &ast::SelfParam,
+    parent_lifetimes: &ast::LifetimeEnv,
+    lookup_id: &LookupId,
+    method_full_path: &ast::Ident, // for better error msg
+    in_path: &ast::Path,
+    env: &Env,
+    errors: &mut Vec<LoweringError>,
+) -> Result<ParamSelf, ()> {
+    match self_param.path_type.resolve(in_path, env) {
+        ast::CustomType::Struct(strct) => {
+            if let Some(tcx_id) = lookup_id.resolve_struct(strct) {
+                if self_param.reference.is_some() {
+                    errors.push(LoweringError::Other(format!("Method `{method_full_path}` takes a reference to a struct as a self parameter, which isn't allowed")));
+                    Err(())
+                } else {
+                    let lifetimes =
+                        TypeLifetimes::from_ast(parent_lifetimes, &self_param.path_type);
+
+                    Ok(ParamSelf::new(SelfType::Struct(StructPath::new(
+                        lifetimes, tcx_id,
+                    ))))
+                }
+            } else if lookup_id.resolve_out_struct(strct).is_some() {
+                if self_param.reference.is_some() {
+                    errors.push(LoweringError::Other(format!("Method `{method_full_path}` takes an out-struct as the self parameter, which isn't allowed. Also, it's also behind a reference, but only opaques can be behind references")));
+                } else {
+                    errors.push(LoweringError::Other(format!("Method `{method_full_path}` takes an out-struct as the self parameter, which isn't allowed")));
+                }
+                Err(())
+            } else {
+                unreachable!(
+                    "struct `{}` wasn't found in the set of structs or out-structs, this is a bug.",
+                    strct.name
+                );
+            }
+        }
+        ast::CustomType::Opaque(opaque) => {
+            let tcx_id = lookup_id.resolve_opaque(opaque).expect("opaque is in env");
+
+            if let Some((lifetime, mutability)) = &self_param.reference {
+                let lifetimes = TypeLifetimes::from_ast(parent_lifetimes, &self_param.path_type);
+                let borrow = Borrow::from_ast(parent_lifetimes, lifetime, *mutability);
+
+                Ok(ParamSelf::new(SelfType::Opaque(OpaquePath::new(
+                    lifetimes,
+                    NonOptional,
+                    borrow,
+                    tcx_id,
+                ))))
+            } else {
+                errors.push(LoweringError::Other(format!("Method `{method_full_path}` takes an opaque by value as the self parameter, but opaques as inputs must be behind refs")));
+                Err(())
+            }
+        }
+        ast::CustomType::Enum(enm) => {
```

Comment:
thought: not entirely sure if we support this well.

noting this now in case this leads up with us needing to write more code when we're hooking it into the JS backend. We can start with unreachable, then.

Reply:
I don't understand, what does JS have to do with this? All of the enums should be in the `lookup_id` because we added all that `Env` gave us. Do you mean we should have `.unwrap_or_else(|| unreachable!("msg"))`?

Reply:
Oh, I was saying that it's _possible_ that our backends currently don't support methods on enums all too well, and in particular with JS we _miiiight_ be completely broken there. And if we are that will become clear when migrating JS over to HIR, in which case we can just unimplemented it on the JS backend side and call it a day.